doc/platform_api/CMIS_and_C-CMIS_support_for_ZR.md

```diff
@@ -0,0 +1,1545 @@
+    ```
+    
+- Example (configure module firmware run):
+    ```
+    admin@sonic:~$ configure interfaces transceiver firmware run Ethernet0
+    Module FW run: Success
+    Module FW run time: 2.00 s
+    ```
+    
+- Example (configure module firmware commit):
+    ```
+    admin@sonic:~$ configure interfaces transceiver firmware commit Ethernet0
+    Module FW commit: Success
+    Module FW commit time: 2.00 s
+    ```
+
+- Example (configure module firmware upgrade, which combines the show FW version, configure download, show download status, configure commit and run and show version in the end):
+    ```
+    admin@sonic:~$ configure interfaces transceiver firmware upgrade qsfp_dd_ver_xx_xx_xx.bin Ethernet0
+    Ethernet0:
+    Start FW downloading
+    Start module FW download: Success
+    Start module FW download time: 0.10 s
+    Total size: 1234567 start bytes: 67 remaining: 1234500
+    Module FW download complete: Success
+    Complete module FW download time: 500.00 s
+    
+    Module FW run: Success
+    Module FW run time: 2.00 s
+    
+    Module FW commit: Success
+    Module FW commit time: 2.00 s
+  
+    Before switch Image A: xxx.xxx.xxx; Run: 1 Commit: 1, Valid: 0
+    Before switch Image B: xxx.xxx.xxx; Run: 0 Commit: 0, Valid: 0
+    Image A: xxx.xxx.xxx; Run: 0 Commit: 0, Valid: 0
+    Image B: xxx.xxx.xxx; Run: 1 Commit: 1, Valid: 0
+    ```
+
+- Example (configure module firmware switch):
+    ```
+    admin@sonic:~$ configure interfaces transceiver firmware switch Ethernet0
+    Module FW run: Success
+    Module FW run time: 2.00 s
+    
+    Module FW commit: Success
+    Module FW commit time: 2.00 s
+  
+    Before switch Image A: xxx.xxx.xxx; Run: 1 Commit: 1, Valid: 0
+    Before switch Image B: xxx.xxx.xxx; Run: 0 Commit: 0, Valid: 0
+    Image A: xxx.xxx.xxx; Run: 0 Commit: 0, Valid: 0
+    Image B: xxx.xxx.xxx; Run: 1 Commit: 1, Valid: 0
+    ```
+
+### 3. High level functions
+
+#### 3.1 Get module basic information
+- get_module_type
+
+```
+def get_module_type(port):
+    module_type = read_reg_from_dict(port, Page00h_Lower.SFF8024_IDENTIFIER)
+    # 400ZR specific: 18h, QSFP-DD
+    return Data_Type_Dict.MODULE_TYPE_DICT[module_type]
+```
+
+- get_module_status
+
+```
+def get_module_status(port):
+    # Bit 3-1:      Module state. 
+    # 000b          -
+    # 001b          ModuleLowPwr
+    # 010b          ModulePwrUp
+    # 011b          ModuleReady This is the only state reported by flat memory modules
+    # 100b          ModulePwrDn
+    # 101b          Fault
+    # 110b          -
+    # 111b          -
+    # Bit 0: Interrupt status. Status of Interrupt output (inverted logic hardware signal)
+    # 0b: Interrupt asserted
+    # 1b: Interrupt not asserted (default)
+    module_status_raw = read_reg_from_dict(port, Page00h_Lower.MODULE_STATE)
+    module_status = (module_status_raw>>1) & 0x7
+    return Data_Type_Dict.MODULE_STATUS_DICT[module_status]
+```
+
+- get_module_vendor
```

Comment:
why do we add new dedicated APIs for getting 'vendor', 'part_numer', and 'serial_number'? they are already part of the "get_transceiver_info"?

Reply:
Modules supporting CMIS (form factors like QSFP-DD, OSFP, etc.) have different memory map and need new method to retrieve this information.